client-spark/common/src/main/java/org/apache/spark/shuffle/ShuffleHandleInfo.java

```diff
@@ -247,4 +330,25 @@ public static ShuffleHandleInfo fromProto(RssProtos.ShuffleHandleInfo handleProt
     handle.remoteStorage = remoteStorageInfo;
     return handle;
   }
+
+  public Set<String> listFaultyServers() {
+    return faultyServerToReplacements.keySet();
+  }
+
+  public void checkPartitionReassignServerNum(
+      Set<Integer> partitionIds, int legalReassignServerNum) {
+    for (int partitionId : partitionIds) {
+      Map<Integer, List<ShuffleServerInfo>> replicas =
+          partitionReplicaAssignedServers.get(partitionId);
+      for (List<ShuffleServerInfo> servers : replicas.values()) {
+        if (servers.size() - 1 > legalReassignServerNum) {
```

Comment:
Is it right? Why do we need `-1`?

Reply:
Yes. The index 0 is the initial assignment server.